### PR TITLE
Add abstract InputStream. Refactor existing FileStreams to in to use it.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michał Zientkiewicz <mzient@gmail.com>

## Category:
**New feature** (*non-breaking change which adds functionality*)
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
An abstract input stream interface is added; existing streams are refactored to inherit from it.

### Tests:

- [ ] Existing tests apply
- [X] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/util/thread_pool.h [hunk 1]
```diff
@@ -34,8 +34,10 @@ class DLL_PUBLIC ThreadPool {
   // Basic unit of work that our threads do
   typedef std::function<void(int)> Work;
 
-  DLL_PUBLIC ThreadPool(int num_thread, int device_id, bool set_affinity,
-                        const std::string &name);
+  DLL_PUBLIC ThreadPool(int num_thread, int device_id, bool set_affinity, const char* name);
+
+  DLL_PUBLIC ThreadPool(int num_thread, int device_id, bool set_affinity, const std::string& name)
+      : ThreadPool(num_thread, device_id, set_affinity, name.c_str()) {}
```
Comment (jantonguirao) [inline dali/pipeline/util/thread_pool.h:40]: Those changes don't belong to this PR, I believe

**Review by jantonguirao — APPROVED** [review]

File: dali/pipeline/util/thread_pool.h (hunk 1 above)
Comment (mzient) [reply dali/pipeline/util/thread_pool.h:40]: This answers why it's signed off as Michal S. - something bad happened when I was rebasing.

File: include/dali/core/stream.h [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  /**
+   * @brief Moves the read pointer in the stream.
+   *
+   * If the new pointer would be out of range, the function may either clamp it to a valid range
+   * or throw an error.
+   *
+   * @param pos     the offset to move
+   * @param whence  the beginning - SEEK_SET, SEEK_CUR or SEEK_END
+   */
+  virtual void SeekRead(ptrdiff_t pos, int whence = SEEK_SET) = 0;
```
Comment (mzient) [inline include/dali/core/stream.h:136]: This is `SeekRead` and not just `Seek` for two reasons:
1. To avoid name collision when we have InputStream and OutputStream and a class that inherits both
2. To emphasize that SeekRead and SeekWrite may be independent (although they don't have to - fseek will move both).

File: include/dali/core/stream.h [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+   */
+  virtual ssize_t TellRead() const = 0;
+
+  /**
+   * @brief Returns the length, in bytes, of the stream.
+   */
+  virtual size_t Size() const = 0;
+
+  /**
+   * @brief Returns the size a signed integer.
```
Comment (stiepan) [inline include/dali/core/stream.h:149]:
```suggestion
   * @brief Returns the size as a signed integer.
```
?
